File: Go/1690_Stone_Game_VII.go
import (
	"fmt"
	"math"
)

func min(nums ...int) int {
	result := math.MaxInt
	for _, num := range nums {
		if result > num {
			result = num
		}
	}
	return result
}

func max(nums ...int) int {
	result := math.MinInt
	for _, num := range nums {
		if result < num {
			result = num
		}
	}
	return result
}

func stoneGameVII(stones []int) int {
  N := len(stones)

	prefSum := make([]int, N+1)
	for i := 0; i < N; i++ {
		prefSum[i+1] = prefSum[i] + stones[i]
	}

	dp := make([][]int, N)
	for i := 0; i < N; i++ {
		dp[i] = make([]int, N)
	}

	score := func(lhs, rhs int) int {
		return prefSum[rhs+1] - prefSum[lhs]
	}

	var memo func(int, int) int
	memo = func(lhs, rhs int) int {
		if lhs == rhs {
			return 0
		}
		if dp[lhs][rhs] != 0 {
			return dp[lhs][rhs]
		}
		result := max(
			score(lhs+1, rhs)-memo(lhs+1, rhs),
			score(lhs, rhs-1)-memo(lhs, rhs-1),
		)
		dp[lhs][rhs] = result
		return result
	}
	return memo(0, N-1)
}

File: CPlusPlus/1690_Stone_Game_VII.cpp
int score(vector<int>& array, int lhs, int rhs) {
    return array[rhs + 1] - array[lhs];
}

int memo(vector<vector<int>>& dp, vector<int>& prefSum, int lhs, int rhs) {
    if (lhs == rhs) return 0;
    if (dp[lhs][rhs] != -1) return dp[lhs][rhs];

    int result = max(
        score(prefSum, lhs + 1, rhs) - memo(dp, prefSum, lhs + 1, rhs),
        score(prefSum, lhs, rhs - 1) - memo(dp, prefSum, lhs, rhs - 1));
    return dp[lhs][rhs] = result;
}

int stoneGameVII(vector<int>& stones) {
    int N = stones.size();
    vector<int> prefSum(N + 1);
	  int count = 1;
    for (auto stone : stones) {
        prefSum[count++] += prefSum[count - 1] + stone;
    }

		// or
		// int count = 0;
    // for (auto stone : stones) {
    //     prefSum[++count] += prefSum[count] + stone;
		// }

    vector<vector<int>> dp(N, vector<int>(N, -1));
    return memo(dp, prefSum, 0, N - 1);
}
